### PR TITLE
feat(cheatcodes): additional cheatcodes to aid in symbolic testing

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation
- #4072 : 
- added `copyStorage(address from, address to)`
- added `setArbitraryStorage(address target)`
- added `mockFunction(address callee, address target, bytes calldata data)`

To be followed up with a PR for adding remaining `expect*Call` and `fresh*` cheatcodes.

## Solution

- `setArbitraryStorage(address target)`:
  - added in `Utilities` group
  - when called, the `target` address is added in `arbitrary_storage` vec
  - in cheatcode `inspector.step_end`, generate random value if target address is marked as arbitrary storage, current opcode is SLOAD and storage slot untouched (cold `ensure_arbitrary_storage` fn added)
  - apply same logic to `vm.load` cheatcode
  - tests and kontrol ported tests added

- `copyStorage(address from, address to)`:
  - added in `Utilities` group
  - when called, from and to accounts are loaded from db and storage of to account replaced with storage of from account
  - tests and kontrol ported tests added

- `mockFunction(address callee, address target, bytes calldata data)`:
  - added in `Evm` group (same with other `mock*` cheatcodes)
  - when called, mapping of `data` -> `target` address is added for `callee` (new hashmap `mocked_functions` added)
  - in stack `inspector.call`, mock target is looked up based on call's calldata (strict match first, if no match then matching by selector is attempted) and call's bytecode address is replaced with mock target, e.g. if using
    ```Solidity
        vm.mockFunction(
            address(my_contract),
            address(model_contract),
            abi.encodeWithSelector(MockFunctionContract.mocked_args_function.selector, 456)
        );
    ```
    then `my_contract.mocked_args_function(456)` will be executed with `model_contract` address bytecode .
    Catch-all function mocks can be set too, for example if using: 
    ```Solidity
        vm.mockFunction(
            address(my_contract),
            address(model_contract),
            abi.encodeWithSelector(MockFunctionContract.mocked_args_function.selector)
        );
    ```
    then `my_contract.mocked_args_function(ANY_VALUE)` will be executed with `model_contract` address bytecode
  - tests and kontrol ported tests added

### CODE REVIEW

File: crates/cheatcodes/tests/random.rs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    function getOwner(uint256 pos) public view returns (address) {
+        return owners[pos];
+    }
+
+    function setOwner(uint256 pos, address owner) public {
+        owners[pos] = owner;
+    }
+}
+
+contract CounterFreshStorageTest is DSTest {
```
Comment (ehildenb) [inline crates/cheatcodes/tests/random.rs:121]: These tests look good overall. I would also add several more tests:

- One that makes the storage fresh, then which reads/writes multiple storage values, and checks that we don't accidentally get aliasing between them somehow.
- One that makes the storage fresh, then reads two different values, and puts a precondition (assume) that they are distinct value (not equal to each other), and then does something with those values.
- One which reads a storage variable, but then bounds the result to something very small. This is to check if we can efficiently generate random numbers within a bound.

File: crates/cheatcodes/tests/random.rs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        console.log(vm.randomUint(min, max));
+        console.log(vm.randomUint(min, max));
+    }
+
+    function testDeterministicRandomAddress() public {
+        console.log(vm.randomAddress());
+        console.log(vm.randomAddress());
+        console.log(vm.randomAddress());
+    }
+}
```
Comment (ehildenb) [inline crates/cheatcodes/tests/random.rs:34]: I guess I would prefer the name `freshUInt`, `freshAddress`, for example, rather than random. Maybe they can be included in the forge-std, but then just simplify down to teh `random` variants? The reason is because we want to use the same cheatcodes as Foundry directly, but for us they're not random, but symbolic. `fresh` captures both random and symbolic as sub-categories.

File: crates/cheatcodes/tests/random.rs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        address owner = counter.getOwner(index);
+        // Subsequent calls should retrieve same value
+        assertEq(counter.getOwner(index), owner);
+        // Change slot and make sure new value retrieved
+        counter.setOwner(index, address(111));
+        assertEq(counter.getOwner(index), address(111));
+
+        require(counter.a() != 0);
+        require(counter.b() != address(0));
+        require(counter.c() != 0);
```
Comment (ehildenb) [inline crates/cheatcodes/tests/random.rs:139]: I actually don't see why these should hold.... As far as I can see, `a`, `b`, and `c` should all be fresh, meaning they _could_ be zero as well as any other number.

Comment (grandizzy) [reply crates/cheatcodes/tests/random.rs:139]: yes, that's correct, they could be 0 though is unlikely. I will remove these assertions as they could fail the test

File: crates/cheatcodes/tests/random.rs (hunk 2 above)
Comment (grandizzy) [reply crates/cheatcodes/tests/random.rs:34]: the `randomAddress` and uint are already added, going to work on the `fresh*` cheatcodes and maybe look into aliasing existing

File: crates/cheatcodes/tests/random.rs (hunk 1 above)
Comment (grandizzy) [reply crates/cheatcodes/tests/random.rs:121]: sounds good, will add such

File: crates/cheatcodes/tests/random.rs (hunk 3 above)
Comment (grandizzy) [reply crates/cheatcodes/tests/random.rs]: @ehildenb I removed the assertions and added a test using seed / producing deterministic values here:
https://github.com/foundry-rs/foundry/blob/46d23223687ea373500ac852e892645eefe2d997/crates/cheatcodes/tests/random.rs#L198-L216

Also ported Kontrol tests using same seed:
https://github.com/foundry-rs/foundry/blob/46d23223687ea373500ac852e892645eefe2d997/crates/cheatcodes/tests/random.rs#L246-L277

File: crates/cheatcodes/tests/random.rs (hunk 1 above)
Comment (grandizzy) [reply crates/cheatcodes/tests/random.rs:123]: added a test here that I *think* it covers first 2 bullets

https://github.com/foundry-rs/foundry/blob/46d23223687ea373500ac852e892645eefe2d997/crates/cheatcodes/tests/random.rs#L152-L165

not clear though on how a test for 3rd one should look like?

File: crates/cheatcodes/tests/random.rs (hunk 3 above)
Comment (ehildenb) [reply crates/cheatcodes/tests/random.rs]: Nice, that's cool.

File: crates/cheatcodes/tests/random.rs (hunk 1 above)
Comment (ehildenb) [reply crates/cheatcodes/tests/random.rs:123]: Hmmm, the third test maybe doesn't make sense, I agree. That one can probably be left for now, and if we encounter unexpected behavior with the tool we can report it and get a test added then.

The main concern is that a user may want to do this:

```
vm.freshStorage(address(counter));
vm.bound(counter.a(), 10, 20);
```

This basically would be saying: set the storage to all symbolic, but the value of `a` should be between `10` and `20`. Currently, I assume this would cause a lot of reverts because the `counter.a()` read will happen before the call to `bound`.

I'm not sure there is anything you can do about this in Foundry, it should work fine in Kontrol. But instead, for Foundry, you would have to do:

```
vm.freshStorage(address(counter));
uint256 a = vm.freshUInt(256, 10, 20); // produces the fresh value _and_ bounds it
counter.setA(a);
```

Then we know that the storage is all symbolic, but the correct bounds exist on `a`.

Comment (ehildenb) [reply crates/cheatcodes/tests/random.rs:123]: The above is psuedocode, btw, just trying to brainstorm how to solve this issue of having a fully symbolic storage, but also allowing some bounds on specific values.

**Review by mattsse — COMMENTED** [review]
love all the additional tests, only have some questions about terminology

File: crates/cheatcodes/spec/src/vm.rs [hunk 4]
```diff
@@ -2303,6 +2312,14 @@ interface Vm {
     /// Unpauses collection of call traces.
     #[cheatcode(group = Utilities)]
     function resumeTracing() external view;
+
+    /// Random storage for given target address.
```
[the remark this answers is not in the document]
Comment (grandizzy) [reply crates/cheatcodes/spec/src/vm.rs:2316]: yeah, agree it's confusing, it's basically random values, in https://github.com/foundry-rs/foundry/issues/4072 there were 2 more suggestions for the name: `symbolicStorage` and `setArbitraryStorage` I personally like the `setArbitraryStorage` one or `randomStorage`, CC @ehildenb for thoughts

[the remark this answers is not in the document]
Comment (mattsse) [reply crates/cheatcodes/spec/src/vm.rs:2316]: I like setArbitraryStorage

File: crates/cheatcodes/tests/assume.rs [hunk 5]
```diff
@@ -0,0 +1,90 @@
+//! Contains various tests for `assume*` cheatcodes.
+
+use alloy_primitives::U256;
+use foundry_config::{Config, FuzzConfig};
+use foundry_test_utils::{forgetest_init, str};
+
+forgetest_init!(test_assume_no_revert, |prj, cmd| {
```
Comment (DaniPopes) [inline crates/cheatcodes/tests/assume.rs:7]: I would like these tests to be all in `forge`

Comment (grandizzy) [reply crates/cheatcodes/tests/assume.rs:7]: yep, will move them all under `forge/tests/it` then

File: crates/forge/tests/cli/test_cmd.rs [hunk 6]
```diff
@@ -137,16 +136,11 @@ contract ATest is DSTest {
     .unwrap();
 
     cmd.arg("test").assert_success().stdout_eq(str![[r#"
-[COMPILING_FILES] with [SOLC_VERSION]
-[SOLC_VERSION] [ELAPSED]
-Compiler run successful!
-
+...
```
Comment (DaniPopes) [inline crates/forge/tests/cli/test_cmd.rs:139]: why was this changed? again the point of all these substitution is so that if the test output changes you can just run the test suite once with SNAPSHOTS=overwrite and it will be updated normally. now you have to do that and also re-add manually `...` since those are manual filters

File: crates/cheatcodes/spec/src/vm.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ehildenb) [reply crates/cheatcodes/spec/src/vm.rs:2316]: Arbitrary sounds good to me too!

[the remark this answers is not in the document]
Comment (ehildenb) [reply crates/cheatcodes/spec/src/vm.rs:2316]: Fresh is just shorter/easier to say.

[the remark this answers is not in the document]
Comment (grandizzy) [reply crates/cheatcodes/spec/src/vm.rs:2316]: kk, will stick with `setArbitraryStorage` then

File: crates/forge/tests/cli/test_cmd.rs (hunk 6 above)
Comment (grandizzy) [reply crates/forge/tests/cli/test_cmd.rs]: yep, reverted, please check (got wrong the `SNAPSHOTS=overwrite` usage but now clear, thank you)

File: crates/cheatcodes/tests/assume.rs (hunk 5 above)
Comment (grandizzy) [reply crates/cheatcodes/tests/assume.rs]: I reverted the change, indeed they should stay in cheats testdata, got 
Will follow up with a PR to cleanup `force/cli/test_cmd.rs` as it is now a catch all thingy

File: crates/cheatcodes/src/inspector.rs [hunk 7]
```diff
@@ -368,6 +371,9 @@ pub struct Cheatcodes {
 
     /// Ignored traces.
     pub ignored_traces: IgnoredTraces,
+
+    /// Addresses that should have fresh storage generated.
```
[the remark this answers is not in the document]
Comment (grandizzy) [reply crates/cheatcodes/src/inspector.rs]: added, pls recheck (did a force push to keep PR clean)

File: crates/cheatcodes/spec/src/vm.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (grandizzy) [reply crates/cheatcodes/spec/src/vm.rs]: changed, pls recheck

File: crates/cheatcodes/src/evm.rs [hunk 8]
```diff
@@ -89,7 +90,13 @@ impl Cheatcode for loadCall {
         let Self { target, slot } = *self;
         ensure_not_precompile!(&target, ccx);
         ccx.ecx.load_account(target)?;
-        let val = ccx.ecx.sload(target, slot.into())?;
+        let mut val = ccx.ecx.sload(target, slot.into())?;
+        // Generate random value if target should have arbitrary storage and storage slot untouched.
+        if ccx.state.arbitrary_storage.contains(&target) && val.is_cold && val.data.is_zero() {
```
Comment (klkvr) [inline crates/cheatcodes/src/evm.rs:95]: what's considered "untouched" here? eg in a single test calling the same contract two times from test scope would result in slot being cold on first call, and warm on second due to how our EVM works. is this expected?

Comment (grandizzy) [reply crates/cheatcodes/src/evm.rs:95]: yep, that's the expected behavior, subsequent loads of same slot should return same value

https://github.com/foundry-rs/foundry/blob/10748dc2d5235a5cef4e8077a3b7bf475e1ef17b/testdata/default/cheats/ArbitraryStorage.t.sol#L108

unless explicitly rewritten

https://github.com/foundry-rs/foundry/blob/10748dc2d5235a5cef4e8077a3b7bf475e1ef17b/testdata/default/cheats/ArbitraryStorage.t.sol#L51

File: crates/cheatcodes/src/inspector.rs [hunk 9]
```diff
@@ -1465,6 +1478,26 @@ impl Cheatcodes {
...
+    fn ensure_arbitrary_storage<DB: DatabaseExt>(
+        &mut self,
+        interpreter: &mut Interpreter,
+        ecx: &mut EvmContext<DB>,
+    ) {
+        if interpreter.current_opcode() == op::SLOAD {
+            let key = try_or_return!(interpreter.stack().peek(0));
+            let target_address = interpreter.contract().target_address;
+            if let Ok(value) = ecx.sload(target_address, key) {
+                if value.is_cold && value.data.is_zero() {
```
Comment (klkvr) [inline crates/cheatcodes/src/inspector.rs:1492]: hmm, I think this would never be cold because revm marks slot as warm during execution of opcode?

File: crates/cheatcodes/src/utils.rs [hunk 10]
```diff
@@ -149,3 +149,25 @@ impl Cheatcode for resumeTracingCall {
...
+        ccx.state.arbitrary_storage.push(*target);
+
+        Ok(Default::default())
+    }
+}
+
+impl Cheatcode for copyStorageCall {
+    fn apply_stateful<DB: DatabaseExt>(&self, ccx: &mut CheatsCtxt<DB>) -> Result {
+        let Self { from, to } = self;
+        if let Ok(from_account) = ccx.load_account(*from) {
```
Comment (klkvr) [inline crates/cheatcodes/src/utils.rs:164]: this won't work for forking but Ig it's out of scope?

File: crates/cheatcodes/src/inspector.rs (hunk 9 above)
Comment (klkvr) [reply crates/cheatcodes/src/inspector.rs:1492]: though it looks like tests are passing? would need to look closer

Comment (klkvr) [reply crates/cheatcodes/src/inspector.rs:1492]: I think it might be because we're reading the value instead of a key in `interpreter.stack().peek(0)`. on `step_end` stack already contains read value

Comment (grandizzy) [reply crates/cheatcodes/src/inspector.rs:1492]: yes, that's my take too, the opcode is not yet executed / slot marked as warm at this point

Comment (klkvr) [reply crates/cheatcodes/src/inspector.rs:1492]: ah so `interpreter.current_opcode()` returns _next_ opcode in `step_end`, so it's fine

Comment (klkvr) [reply crates/cheatcodes/src/inspector.rs:1492]: > yes, that's my take too, the opcode is not yet executed / slot marked as warm at this point

I was thinking that we're entering `ensure_arbitrary_storage` _after_ revm executed the opcode, so it's my bad

though can we document what's happening here anyway? especially that this would happen before sload even though we are in `_end` hook

Comment (grandizzy) [reply crates/cheatcodes/src/inspector.rs:1492]: sure, will add comment to make this clear

**Comment by ehildenb** [comment]
Another test I thought of: basically how do `abitraryStorage` and `copyStorage` interact?

I would expect (and what we would have for symbolic execution) that if I copy a given contracts storage, I get the _same_ storage as that contract. So if I set a storage to arbitrary, then copy it, all future reads from both storages will agree (even if reading from uninitialized slots). So a test could be:

- Create  contract `A` and set its storage to `abitrary`.
- Use `copyStorage` to copy storage of `A` to `B`.
- Read from slots `m` and `n` from both contract `A` and `B`, and assert that they are equal.
- Write to slot `m` of `A`, and assert that now it is different than what is in slot `m` on `B`.

**Comment by grandizzy** [comment]
> Another test I thought of: basically how do `abitraryStorage` and `copyStorage` interact?
> 
> I would expect (and what we would have for symbolic execution) that if I copy a given contracts storage, I get the _same_ storage as that contract. So if I set a storage to arbitrary, then copy it, all future reads from both storages will agree (even if reading from uninitialized slots). So a test could be:
> 
> * Create  contract `A` and set its storage to `abitrary`.
> * Use `copyStorage` to copy storage of `A` to `B`.
> * Read from slots `m` and `n` from both contract `A` and `B`, and assert that they are equal.
> * Write to slot `m` of `A`, and assert that now it is different than what is in slot `m` on `B`.

yeah, this scenario wasn't covered, working to add such

File: crates/cheatcodes/src/utils.rs (hunk 10 above)
Comment (grandizzy) [reply crates/cheatcodes/src/utils.rs:168]: @klkvr please check https://github.com/foundry-rs/foundry/pull/8807/commits/16f068f6e9f557571464b9ad5d39d9f24ff40bb5 Had to accommodate copy from arbitrary storage scenario pointed here https://github.com/foundry-rs/foundry/pull/8807#issuecomment-2339014187 too, so a little bit bigger change, ptal. thanks!

File: crates/cheatcodes/src/utils.rs [hunk 11]
```diff
@@ -149,3 +149,30 @@ impl Cheatcode for resumeTracingCall {
         Ok(Default::default())
     }
 }
+impl Cheatcode for setArbitraryStorageCall {
```
Comment (DaniPopes) [inline crates/cheatcodes/src/utils.rs:152]:
```suggestion

impl Cheatcode for setArbitraryStorageCall {
```

File: crates/forge/tests/it/cheats.rs [hunk 12]
```diff
@@ -26,6 +27,7 @@ async fn test_cheats_local(test_data: &ForgeTestData) {
     }
 
     let mut config = test_data.config.clone();
+    config.fuzz.seed = Some(U256::from(100));
```
Comment (DaniPopes) [inline crates/forge/tests/it/cheats.rs:30]: should this be set in test_data?

File: crates/evm/evm/src/inspectors/stack.rs [hunk 13]
```diff
@@ -726,6 +726,18 @@ impl<'a, DB: DatabaseExt> Inspector<DB> for InspectorStackRefMut<'a> {
 
         ecx.journaled_state.depth += self.in_inner_context as usize;
         if let Some(cheatcodes) = self.cheatcodes.as_deref_mut() {
+            // Handle mocked functions, replace bytecode address with mock if matched.
+            if let Some(mocks) = cheatcodes.mocked_functions.get(&call.target_address) {
+                if let Some(target) = mocks.get(&call.input) {
+                    call.bytecode_address = *target;
+                } else {
+                    // Check if we have a catch-all mock set for selector.
+                    if let Some(target) = mocks.get(&call.input.slice(..4)) {
```
Comment (DaniPopes) [inline crates/evm/evm/src/inspectors/stack.rs:735]: can collapse into one `get().or_else(|| )`, also this panics if `call.input.len() < 4`, use `call.input.get(..4)`

File: crates/cheatcodes/src/inspector.rs [hunk 14]
```diff
@@ -1465,6 +1567,63 @@ impl Cheatcodes {
...
+    /// If current opcode to execute is `SLOAD` and storage slot is cold, then an arbitrary value
+    /// is generated and saved in target address storage (therefore when `SLOAD` opcode is executed,
+    /// the arbitrary value will be returned).
+    #[cold]
+    fn ensure_arbitrary_storage<DB: DatabaseExt>(
+        &mut self,
+        interpreter: &mut Interpreter,
+        ecx: &mut EvmContext<DB>,
+    ) {
+        if interpreter.current_opcode() == op::SLOAD {
```
Comment (DaniPopes) [inline crates/cheatcodes/src/inspector.rs:1581]: should move the `SLOAD` check outside into `step_end`, and make it call one `arbitrary_storage_end` to avoid callsite bloat

File: crates/cheatcodes/src/evm.rs [hunk 15]
```diff
@@ -89,7 +90,24 @@ impl Cheatcode for loadCall {
         let Self { target, slot } = *self;
         ensure_not_precompile!(&target, ccx);
         ccx.ecx.load_account(target)?;
-        let val = ccx.ecx.sload(target, slot.into())?;
+        let mut val = ccx.ecx.sload(target, slot.into())?;
+
+        if val.is_cold && val.data.is_zero() {
+            let rand_value = ccx.state.rng().gen();
```
Comment (DaniPopes) [inline crates/cheatcodes/src/evm.rs:96]: should not gen unless either is true

File: crates/forge/tests/it/cheats.rs (hunk 12 above)
Comment (grandizzy) [reply crates/forge/tests/it/cheats.rs]: good point, not sure we should be adding a new `TEST_DATA_WITH_SEED` (we don't have one for isolate either) but default tests shouldn't run with seed - since there's no option to set it inline I've added a test that runs contracts matching `WithSeed` (excluded from defaults), hope this makes sense
pls check https://github.com/foundry-rs/foundry/pull/8807/commits/7e69723609d13ceb3ce744c0f7bbdd3e156f1296

File: crates/cheatcodes/src/evm.rs (hunk 15 above)
Comment (grandizzy) [reply crates/cheatcodes/src/evm.rs]: yep, changed in https://github.com/foundry-rs/foundry/pull/8807/commits/7e69723609d13ceb3ce744c0f7bbdd3e156f1296

File: crates/cheatcodes/src/inspector.rs (hunk 14 above)
Comment (grandizzy) [reply crates/cheatcodes/src/inspector.rs]: changed in https://github.com/foundry-rs/foundry/pull/8807/commits/7e69723609d13ceb3ce744c0f7bbdd3e156f1296

File: crates/evm/evm/src/inspectors/stack.rs (hunk 13 above)
Comment (grandizzy) [reply crates/evm/evm/src/inspectors/stack.rs]: changed in https://github.com/foundry-rs/foundry/pull/8807/commits/7e69723609d13ceb3ce744c0f7bbdd3e156f1296

File: crates/forge/tests/it/cheats.rs [hunk 16]
```diff
@@ -7,14 +7,16 @@ use crate::{
...
 use foundry_test_utils::Filter;
 
-/// Executes all cheat code tests but not fork cheat codes or tests that require isolation mode
+/// Executes all cheat code tests but not fork cheat codes or tests that require isolation mode or
+/// specific seed.
 async fn test_cheats_local(test_data: &ForgeTestData) {
     let mut filter = Filter::new(".*", ".*", &format!(".*cheats{RE_PATH_SEPARATOR}*"))
         .exclude_paths("Fork")
-        .exclude_contracts("Isolated");
+        .exclude_contracts("(Isolated|WithSeed)");
```
Comment (klkvr) [inline crates/forge/tests/it/cheats.rs:19]: please add it to `if cfg!(feature = "isolate-by-default")` branch below too, tests with `--features=isolate-by-default` are failing because we're overriding it there

Comment (grandizzy) [reply crates/forge/tests/it/cheats.rs:19]: sorry, missed running the test with feature set locally, added here https://github.com/foundry-rs/foundry/pull/8807/commits/02086f94db6b6db262ded77b53d5c480f9f4fc8a (had to add mock function test too)

**Review by klkvr — APPROVED** [review]
LGTM

File: crates/forge/tests/it/cheats.rs (hunk 16 above)
Comment (klkvr) [reply crates/forge/tests/it/cheats.rs:19]: > had to add mock function test too

hmm, wondering why would it break with `--isolate`?

Comment (klkvr) [reply crates/forge/tests/it/cheats.rs:19]: ah, I see, it's because we're ignoring `bytecode_address` when handling isolated calls https://github.com/foundry-rs/foundry/blob/2c73013a01264e8577e728bee33961845d126963/crates/evm/evm/src/inspectors/stack.rs#L745

Comment (klkvr) [reply crates/forge/tests/it/cheats.rs:19]: not sure how to handle this now, probably not a big deal

could be solved by instead of altering `bytecode_address` changing target address bytecode for duration of the call, though it would introduce new edge cases

Comment (grandizzy) [reply crates/forge/tests/it/cheats.rs:19]: got it, thanks for clarifying! do you think we should be looking into supporting it as well or good to have it only for non isolating calls for now?

File: crates/cheatcodes/src/inspector.rs [hunk 17]
```diff
@@ -1045,14 +1136,22 @@ impl<DB: DatabaseExt> Inspector<DB> for Cheatcodes {
...
         }
 
         if self.gas_metering.touched {
             self.meter_gas_check(interpreter);
         }
+
+        // `setArbitraryStorage` and `copyStorage`: add arbitrary values to storage.
+        if (self.arbitrary_storage.is_arbitrary(&interpreter.contract().target_address) ||
+            self.arbitrary_storage.is_copy(&interpreter.contract().target_address)) &&
+            interpreter.current_opcode() == op::SLOAD
```
Comment (DaniPopes) [inline crates/cheatcodes/src/inspector.rs:1151]: this should be `op::SLOAD` check, please be very mindful of what goes in step and step_end because these functions are called millions of times

Comment (grandizzy) [reply crates/cheatcodes/src/inspector.rs:1151]: oh, ok, will do a PR to change this, mind to explain the difference? (was thinking that is better to check if is arbitrary or copy of arbitrary storage first and only after if one of them true if it's a SLOAD, as arbitrary and copy are not so commons)

Comment (DaniPopes) [reply crates/cheatcodes/src/inspector.rs:1151]: They are called on execution of every single opcode, meaning tens to hundreds of thousands per single execution of a test; we want them inlined and optimized for the common path of "do nothing" as most of these checks will be false most of the time

Comment (DaniPopes) [reply crates/cheatcodes/src/inspector.rs:1151]: wrt the checks order, the thinking is not correct because checking opcode is 1 or 2 instructions, while the others are hashmap lookups which are in the order of hundreds or thousands

Comment (grandizzy) [reply crates/cheatcodes/src/inspector.rs:1151]: thank you, makes total sense, will follow up with a PR to get it inline

Comment (grandizzy) [reply crates/cheatcodes/src/inspector.rs:1151]: wonder if better to have arbitrary storage as an option, made a draft PR here https://github.com/foundry-rs/foundry/pull/8848
